feat: Provide support for property-based authorization management

ce-dmelnych opened the pull request:
## Description

This PR introduces the foundation for property-based authorization support, extending the authorization beyond resource ID-based and wildcard (`*`) permissions.
Property-based authorizations allow defining authorizations that grant permissions based on specific resource properties (such as `assignee`, `candidateUsers`, `candidateGroups` for user tasks), enabling more flexible and dynamic access control without requiring individual resource IDs.

**How it will work:** Property-based authorizations will be checked based on resource property values rather than explicit resource identifiers. For example, an authorization can be configured to grant a user permission to access any user task where they are the assignee, automatically applying to all matching tasks without needing to specify each task ID individually. This enables organizations to define authorization rules that dynamically adapt as task assignments change.

**What this PR includes:**
- New `PROPERTY` matcher, added to `AuthorizationResourceMatcher`
- Introduced `resourcePropertyName` field to `AuthorizationRecord` and `AuthorizationScope` to specify which property the authorization evaluates (e.g., "assignee", "candidateUsers")
- Verifies the ability to create, update, and delete property-based authorizations through the Engine
- Enforced mutual exclusivity between `resourceId` and `resourcePropertyName` fields in `AuthorizationCreateProcessor`, `AuthorizationUpdateProcessor` - exactly one must be provided, rejecting requests with neither or both

**Note:** This PR focuses solely on providing the support to create, update, delete property-based authorizations in the Engine. The evaluation logic that actually checks property-based permissions during runtime authorization decisions will be implemented in a following PRs.

## Related issues

closes #40016

ce-dmelnych committed:
feat: add PROPERTY matcher and property name to AuthorizationScope

ce-dmelnych committed:
feat: add resource property name to AuthorizationRecord

ce-dmelnych committed:
feat: add resource property name to PersistedAuthorization

ce-dmelnych committed:
feat: add validation for resource property name on authorization creation

ce-dmelnych committed:
test: add resourcePropertyName field to JSON serialization tests

ce-dmelnych committed:
test: add resourcePropertyName to AuthorizationRecord golden file

Introduced the new `resourcePropertyName` field in AuthorizationRecord with a default
empty-string value. This is a backward-compatible change, so updating the golden file
is safe.

ce-dmelnych committed:
fix: increase segment count in shouldNotReceiveEntriesOnDisconnect

The test failure (4 < 4) indicates the disconnected broker now reaches
the same segment count as the threshold. This is likely due to increased
record sizes from adding the `resourcePropertyName` field to authorization
models, affecting segment creation during test setup.

Increase the threshold to 5 to maintain a buffer between the disconnected
broker's segments and the expected count, ensuring the test verifies
disconnect behavior.

ce-dmelnych committed:
test: add tests for property-based auth deletion and updates

ce-dmelnych committed:
feat: add validation for authorization updates

ce-dmelnych committed:
feat: display property name in authorization logs

Property-based authorizations now log as:
  USER "ownerId" can [COMPLETE] USER_TASK based on "assignee" property

ce-dmelnych committed:
refactor: use switch for resource matcher validation

Replace if-else with switch expression in AuthorizationCreate/UpdateProcessor
for cleaner handling of PROPERTY, ID, and ANY matcher cases.

ce-dmelnych committed:
feat: add default case to authorization matcher validation

Add a default no-op case to the `validateResourceIdOrPropertyName` method
to maintain backward compatibility.
Some existing tests are using UNSPECIFIED as a value for matcher
(e.g., AuthorizationUpdateAuthorizationTest#shouldBeAuthorizedToUpdateAuthorizationWithDefaultUser).

The validation now only enforces rules for PROPERTY, ID, and ANY matchers,
allowing UNSPECIFIED and future matcher types to pass.

ce-dmelnych commented on zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/cluster/clustering/FailOverReplicationTest.java:
```diff
@@ -77,7 +77,7 @@ public void shouldFailOverOnNetworkPartition() {
   public void shouldNotReceiveEntriesOnDisconnect() {
     // given
-    final var segmentCount = 4;
+    final var segmentCount = 5;
```
I've adjusted this value because seems like adding the `resourcePropertyName` property to `AuthorizationRecord`,`PersistedAuthorization` models increased their size, causing one additional segment to be created during test setup.

@npepinpe - You recently bumped this from 2 to 4 for similar reasons. Could you confirm:
1. Is this increase expected behavior when adding properties to records?
2. Should we be concerned about record size growth affecting segment creation patterns?
3. Would you recommend any alternative approach to handle this in tests?

This value remained at 2 for ~5 years before your change, so wanted to verify this pattern is acceptable or if we should investigate the underlying segment sizing.

mustafadagher commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/AuthorizationUpdateProcessor.java:
```diff
@@ -123,4 +127,56 @@ private void writeEventAndDistribute(
+      }
+      default -> {
+        // No additional validation for other matchers to keep backward compatibility
+      }
```
I agree to option 2 as well, but I think someone from identity team would be a better candidate to discuss this decision

npepinpe replied:
Ben's welcome to give his thoughts, but I will review this PR atm, as he's stretched a little too thin :)

npepinpe replied:
Definitely option 2. Convenience for tests is not a strong enough argument in this instance, considering the downsides.

mustafadagher commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/AuthorizationUpdateProcessor.java:
```diff
@@ -138,37 +137,40 @@ private Either<Rejection, AuthorizationRecord> validateResourceIdOrPropertyName(
-            new Rejection(
-                RejectionType.INVALID_ARGUMENT,
-                "Expected to update authorization with matcher 'PROPERTY', but no resource property name was provided. Please provide a resource property name."));
+    switch (matcher) {
```
🔧 does it make sense to extract those validations outside and re-use them. I know there're not so many places to re-use, it's only those two locations. But they could be easily updated in one location and be forgotten in the other in the future.

I see also they differ a tiny bit on the rejection message. 

WDYT?

npepinpe replied:
I think extracting and reusing the validation makes sense, even if the error messages won't be ultra specific. Ultimately it's validating the resulting entity (e.g. mutually exclusive property/id based on type, non-null, etc.).

I'm fine with it as a follow up as long as we commit to doing it fairly quickly (< 2 weeks), before we build more on top of this duplication. Otherwise let's do it now.

mustafadagher reviewed (approved):
Thanks @ce-dmelnych , great work! 🚀 
🔧 I only left a suggestion to refactor the validations out, feel free to take it or leave it.

npepinpe commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/AuthorizationCreateProcessor.java:
```diff
@@ -112,4 +116,56 @@ private void writeEventAndDistribute(
+                      "Expected to create authorization with matcher '%s', but a resource property name was provided. Resource property names are only valid for matcher 'PROPERTY'.",
+                      matcher)));
+        }
+        if (!hasResourceId) {
```
🤔 Do I understand it right that there was previously no validation for this?

npepinpe commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/authorization/AuthorizationCreateAuthorizationTest.java:
```diff
@@ -128,6 +129,89 @@ public void shouldBeForbiddenToCreateAuthorizationIfNoPermissions() {
   }
 
+  @Test
+  public void shouldBeAuthorizedToCreatePropertyBasedAuthorizationWithDefaultUser() {
```
❓ Can you explain the value of this and the test below? Permissions are not so granular that we can give, say, authorization to create authorizations with a property or ID matcher. You can either create or not create authorizations, regardless of the resource matcher. Or did I misunderstand?

npepinpe commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/authorization/CreateAuthorizationMultipartitionTest.java:
```diff
@@ -193,4 +193,83 @@ public void shouldRejectAuthorizationWithNonexistentMappingInMultiPartitionSetup
+            tuple(CommandDistributionIntent.STARTED, RecordType.EVENT))
+        .endsWith(tuple(CommandDistributionIntent.FINISHED, RecordType.EVENT));
+
+    // Verify authorization is created on all partitions
```
```suggestion
    // Verify authorization is created on all other partitions, as we already verified it above for partition 1
```

npepinpe reviewed (approved):
Good stuff, please address all comments before merging though, even if it's there are no changes required :)

ce-dmelnych committed:
refactor: move resource matcher check to AuthorizationEntityChecker

Move validateResourceMatcher logic from AuthorizationCreateProcessor and
AuthorizationUpdateProcessor to AuthorizationEntityChecker. This eliminates
duplicate code across the two processors and centralizes validation logic in a single place.

ce-dmelnych committed:
test: remove duplicate resource matcher validation tests

Remove redundant test cases from CreateAuthorizationTest and
UpdateAuthorizationTest that verify resource matcher validation logic.
This validation is now comprehensively tested in AuthorizationEntityCheckerTest
covering all scenarios.

Keep one test case in each processor test class to ensure the validation
is still called from the processors and prevent accidental removal of
validateResourceMatcher invocations.

ce-dmelnych committed:
feat: reject authorization operations with UNSPECIFIED matcher

Add validation to reject authorization create and update operations
when the resource matcher is UNSPECIFIED. This ensures that all authorizations
have a valid matcher type (ID, ANY, or PROPERTY).

ce-dmelnych committed:
test: replace UNSPECIFIED matcher usage with valid matchers in tests

UNSPECIFIED matcher was previously used in tests for simplicity when testing
other aspects of authorization. Now that validation is enforced, all tests
properly specify valid matcher types.

ce-dmelnych committed:
feat: enforce explicit matcher validation with default case

Add default case to `validateResourceMatcher` switch to ensure any new
AuthorizationResourceMatcher values must be explicitly handled with
proper validation logic rather than being silently ignored.

ce-dmelnych commented:
Hi @npepinpe! 👋

I've addressed all your feedback. Could you please re-review the PR, new commits being added starting from the [commit](https://github.com/camunda/camunda/pull/40191/commits/d58948e85fad8cda1bfcdce3c6aa74d5f3dde9af)?

Thanks for the thorough review! 🙏
